### PR TITLE
deployment.keys:  'keyFile' option to provide secret contents from local files

### PR DESCRIPTION
**Pull request by deepfire**
This is supposed to address https://github.com/NixOS/nix/issues/1307 by adding an option `filename` (as an alternative to `text`) to `deployments.keys` to avoid ever storing a secret in Nix string.

Note, that this doesn't add the corresponding part to `system.activationScripts.nixops-keys`.

### CODE REVIEW

File: nixops/backends/__init__.py [hunk 1]
```diff
@@ -212,7 +212,15 @@ def send_keys(self):
         for k, opts in self.get_keys().items():
             self.log("uploading key ‘{0}’...".format(k))
             tmp = self.depl.tempdir + "/key-" + self.name
-            f = open(tmp, "w+"); f.write(opts['text']); f.close()
+
+            # Ensure that keys.<key>.file XOR keys.<key>.file
+            self.log ("'text' in opts: %s, 'filename' in opts: %s" % ('text' in opts, 'filename' in opts))
+            assert     ('text' in opts  or 'filename' in opts), "Specification error for key '%s': neither `keys.%s.text` nor `keys.%s.file` was specified."  % (k, k, k)
+            assert not ('text' in opts and 'filename' in opts), "Specification error for key '%s': `keys.%s.text` and `keys.%s.file` are mutually exclusive." % (k, k, k)
```
Comment (ip1981) [inline nixops/backends/__init__.py]: Isn't assert in Nix more explicit?

**Comment by ip1981** [comment]
Almost the same can be achieved with `builtins.readFile`. I mean not in NixOps, but in machines specifications.

**Comment by domenkozar** [comment]
@ip1981 except that with `builtins.readFile` you hit https://github.com/NixOS/nix/issues/1307 as random data can contain null bytes.

File: nix/keys.nix [hunk 2]
```diff
@@ -6,13 +6,35 @@ let
...
+      '';
+    };
+
+    options.filename = mkOption {
+      example = "./super-secret.key";
+      default = null;
+      type = types.nullOr types.str;
+      description = ''
+
+        When this is non-null, contents of the specified file will be deployed to
```
Comment (domenkozar) [inline nix/keys.nix]: s/file/filename/

File: nix/keys.nix [hunk 3]
```diff
@@ -6,13 +6,35 @@ let
...
+        specified.
+      '';
+    };
+
+    options.filename = mkOption {
+      example = "./super-secret.key";
+      default = null;
+      type = types.nullOr types.str;
+      description = ''
+
```
Comment (domenkozar) [inline nix/keys.nix]: (nitpick) Extra unneeded empty line.

File: nix/keys.nix [hunk 4]
```diff
@@ -6,13 +6,35 @@ let
...
-        The text the key should contain. So if the key name is
-        <replaceable>password</replaceable> and <literal>foobar</literal>
-        is set here, the contents of the file
+        When this is non-null, this designates the text that the key should
+        contain. So if the key name is <replaceable>password</replaceable>
+        and <literal>foobar</literal> is set here, the contents of the file
         <filename>/run/keys/<replaceable>password</replaceable></filename>
         will be <literal>foobar</literal>.
+
+        NOTE: either this XOR the <literal>filename</literal> option ought to be
```
Comment (domenkozar) [inline nix/keys.nix]: I'd rephrase this to: `NOTE: Either <literal>text</literal> or <literal>filename</literal> have to be set.`

File: nix/keys.nix [hunk 5]
```diff
@@ -6,13 +6,35 @@ let
...
+      description = ''
+
+        When this is non-null, contents of the specified file will be deployed to
+        the specified key on the target machine.  If the key name is
+        <replaceable>password</replaceable> and <literal>/foo/bar</literal> is set
+        here, the contents of the file
+        <filename>/run/keys/<replaceable>password</replaceable></filename> will be
+        the same as that of <literal>/foo/bar</literal> on the deploying host.
+
+        NOTE: either this XOR the <literal>text</literal> option ought to be
```
Comment (domenkozar) [inline nix/keys.nix]: I'd rephrase this to: NOTE: Either <literal>text</literal> or <literal>filename</literal> have to be set.

File: nix/keys.nix [hunk 6]
```diff
@@ -6,13 +6,35 @@ let
...
+      example = "./super-secret.key";
+      default = null;
+      type = types.nullOr types.str;
+      description = ''
+
+        When this is non-null, contents of the specified file will be deployed to
+        the specified key on the target machine.  If the key name is
+        <replaceable>password</replaceable> and <literal>/foo/bar</literal> is set
+        here, the contents of the file
+        <filename>/run/keys/<replaceable>password</replaceable></filename> will be
```
Comment (domenkozar) [inline nix/keys.nix]: the contents of the file deployed ... will be the same as local file <literal>/foo/bar</literal>

File: nix/keys.nix [hunk 7]
```diff
@@ -133,6 +155,8 @@ in
               let
                 # FIXME: The key file should be marked as private once
                 # https://github.com/NixOS/nix/issues/8 is fixed.
+                #
+                # FIXME: implement value.filename for this as well
```
Comment (domenkozar) [inline nix/keys.nix]: I suggest we fail here if `text` is not set.

Given the discussion in https://github.com/NixOS/nixops/pull/650, this will change soon so no need to implement `filename` handling sanely.

File: nixops/backends/__init__.py [hunk 8]
```diff
@@ -212,7 +212,14 @@ def send_keys(self):
         for k, opts in self.get_keys().items():
             self.log("uploading key ‘{0}’...".format(k))
             tmp = self.depl.tempdir + "/key-" + self.name
-            f = open(tmp, "w+"); f.write(opts['text']); f.close()
+
+            # Ensure that keys.<key>.text XOR keys.<key>.filename
+            assert     ('text' in opts  or 'filename' in opts), "Specification error for key '%s': neither `keys.%s.text` nor `keys.%s.filename` was specified."  % (k, k, k)
+            assert not ('text' in opts and 'filename' in opts), "Specification error for key '%s': `keys.%s.text` and `keys.%s.filename` are mutually exclusive." % (k, k, k)
```
Comment (domenkozar) [inline nixops/backends/__init__.py]: Let's move this to Nix. For example like in https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/config/users-groups.nix#L539-L564

File: nixops/backends/__init__.py [hunk 9]
```diff
@@ -212,7 +212,14 @@ def send_keys(self):
...
+
+            if   'text'     in opts: f = open(tmp, "w+"); f.write(opts['text']); f.close()
+            elif 'filename' in opts: self._logged_exec(["cp", opts['filename'], tmp])
```
Comment (domenkozar) [inline nixops/backends/__init__.py]: A bit more Pythonic code would be:

```python
    if 'text' in opts: 
        with open(tmp, "w+") as f:
            f.write(opts['text'])
    elif 'filename' in opts:
         self._logged_exec(["cp", opts['filename'], tmp])
    else:
         raise Exception("Neither text of filename were set ...")
```

File: nix/keys.nix [hunk 10]
```diff
@@ -6,13 +6,35 @@ let
...
     options.text = mkOption {
       example = "super secret stuff";
-      type = types.str;
+      default = null;
+      type = types.nullOr types.str;
       description = ''
-        The text the key should contain. So if the key name is
-        <replaceable>password</replaceable> and <literal>foobar</literal>
-        is set here, the contents of the file
+        When this is non-null, this designates the text that the key should
```
Comment (domenkozar) [inline nix/keys.nix]: `When non-null this designates ...`

**Comment by ip1981** [comment]
On the other side, file-based keys are more feasible (to avoid secrets in nix expressions) and should be implemented this way from the start :)

File: nix/keys.nix (hunk 2 above)
Comment (deepfire) [inline nix/keys.nix]: A file name specifies a file, I'd say..

File: nixops/backends/__init__.py (hunk 1 above)
Comment (deepfire) [inline nixops/backends/__init__.py]: Done, moved to Nix option-specific assertions, as per @domenkozar  suggestion.

File: nix/keys.nix (hunk 7 above)
Comment (deepfire) [inline nix/keys.nix]: This will fail (by the virtue of the assertions now moved to Nix) if _neither_ `text` nor `filename` is set is that good enough?

File: nixops/backends/__init__.py (hunk 8 above)
Comment (deepfire) [inline nixops/backends/__init__.py]: Done.

File: nix/keys.nix (hunk 7 above)
Comment (domenkozar) [inline nix/keys.nix]: If only `filename` is set and `deployment.storeKeysOnMachine` is set to true, this will fail since `value.text` one line below will be `null`.

We have to add another assertion to say that `deployment.storeKeysOnMachine && filename` is not supported.

File: nix/keys.nix [hunk 11]
```diff
@@ -6,13 +6,34 @@ let
...
+
+        NOTE: Either <literal>text</literal> or <literal>filename</literal> have
+        to be set.
+      '';
+    };
+
+    options.filename = mkOption {
+      example = "./super-secret.key";
+      default = null;
+      type = types.nullOr types.str;
```
Comment (ip1981) [inline nix/keys.nix]: I'd make it a path, to be sure nix can find the file (by absolute path) :)
It does not need to be a derivation,

File: nixops/backends/__init__.py [hunk 12]
```diff
@@ -212,7 +212,10 @@ def send_keys(self):
         for k, opts in self.get_keys().items():
             self.log("uploading key ‘{0}’...".format(k))
             tmp = self.depl.tempdir + "/key-" + self.name
-            f = open(tmp, "w+"); f.write(opts['text']); f.close()
+
+            if   'text'     in opts: f = open(tmp, "w+"); f.write(opts['text']); f.close()
+            elif 'filename' in opts: self._logged_exec(["cp", opts['filename'], tmp])
```
Comment (ip1981) [inline nixops/backends/__init__.py]: Absolute path would help for files named like `--help` ;)

File: nixops/backends/__init__.py [hunk 13]
```diff
@@ -212,7 +212,15 @@ def send_keys(self):
...
             tmp = self.depl.tempdir + "/key-" + self.name
-            f = open(tmp, "w+"); f.write(opts['text']); f.close()
+
+            if 'text' in opts: 
+                with open(tmp, "w+") as f:
+                    f.write(opts['text'])
+            elif 'filename' in opts:
+                self._logged_exec(["cp", opts['filename'], tmp])
+            else:
+                raise Exception("Neither 'text' or 'filename' options were set for key '%s'." % k)
```
Comment (domenkozar) [inline nixops/backends/__init__.py]: To support Python 3 use "".format()

File: nix/keys.nix (hunk 11 above)
Comment (deepfire) [inline nix/keys.nix]: I have tried to make it a path, but couldn't make the XML query work.

File: nix/keys.nix [hunk 14]
```diff
@@ -6,13 +6,34 @@ let
...
+      description = ''
+        When non-null, contents of the specified file will be deployed to
+        the specified key on the target machine.  If the key name is
+        <replaceable>password</replaceable> and <literal>/foo/bar</literal> is set
+        here, the contents of the file
+        <filename>/run/keys/<replaceable>password</replaceable></filename>
+        deployed will be the same as local file <literal>/foo/bar</literal>.
```
Comment (domenkozar) [inline nix/keys.nix]: (nitpick): explain how this option mitigates https://github.com/NixOS/nix/issues/1307

File: nixops/backends/__init__.py (hunk 12 above)
Comment (ip1981) [inline nixops/backends/__init__.py]: Or use Python's [shutils](https://docs.python.org/3/library/shutil.html) or something like that. Absolute paths are still good :)

File: nix/keys.nix (hunk 11 above)
Comment (ip1981) [inline nix/keys.nix]: If the problem is with slashes, it should be fixed anyway, because someone might wan to use subdirectories: `./secrets/my.key`

File: nixops/backends/__init__.py (hunk 12 above)
Comment (domenkozar) [inline nixops/backends/__init__.py]: I don't understand your comment, what absolute path?

Comment (ip1981) [inline nixops/backends/__init__.py]: From here https://github.com/NixOS/nixops/pull/661#discussion_r115726566. Absolute paths (starting with `/`) could help avoid problems with `cp`.

File: nix/keys.nix (hunk 7 above)
Comment (deepfire) [inline nix/keys.nix]: Done.

File: nix/keys.nix (hunk 11 above)
Comment (deepfire) [inline nix/keys.nix]: No, it's with XML queries for non-`str` types being harder, and unprecedented (according to my, admittedly limited, grep skill) in the nixops repository.  I could get down to the bottom of matter, but it would take more effort.

File: nix/keys.nix [hunk 15]
```diff
@@ -123,6 +144,16 @@ in
       "things to break."
     )];
 
+    assertions = flip mapAttrsToList config.deployment.keys (key: opts: {
+      assertion = (opts.text == null && opts.filename != null) ||
+                  (opts.text != null && opts.filename == null);
+      message = "Deployment key '${key}' must have either a 'text' or a 'filename' specified.";
```
Comment (ip1981) [inline nix/keys.nix]: Add
```
 || (opts.text == null && opts.filename == null)
```

and corresponding python code can be removed.

File: nix/keys.nix (hunk 14 above)
Comment (deepfire) [inline nix/keys.nix]: Done.

File: nix/keys.nix [hunk 16]
```diff
@@ -123,6 +148,16 @@ in
...
+    }) ++        flip mapAttrsToList config.deployment.keys (key: opts: {
+      assertion = !config.deployment.storeKeysOnMachine ||
+                  !opts.filename;
+      message = "On-machine storage of 'filename'-specified deployment keys is not supported. Offending key is: '${key}'.";
```
Comment (domenkozar) [inline nix/keys.nix]: I'd change the wording to: `deployment.storeKeysOnMachine` and `filename`-specified deployment keys is not ...

Comment (ip1981) [inline nix/keys.nix]: But it could work if you capture it into Nix store (`"${filename}"`).

Comment (ip1981) [inline nix/keys.nix]: Not supporting `filename` in some cases is surprising.

File: nix/keys.nix (hunk 11 above)
Comment (deepfire) [inline nix/keys.nix]: Alright, done the switch to `path`-typing.

File: nixops/backends/__init__.py (hunk 12 above)
Comment (deepfire) [inline nixops/backends/__init__.py]: Switched to `pathname` type, so this should be covered, right @ip1981?

File: nixops/backends/__init__.py (hunk 9 above)
Comment (deepfire) [inline nixops/backends/__init__.py]: Done.

**Comment by deepfire** [comment]
@ip1981, do you think we can move this forward without having the `filename` support the `storeKeysOnMachine` case?

Do you have any other concerns?

**Comment by domenkozar** [comment]
cc @rbvermaa @edolstra

**Comment by ip1981** [comment]
> do you think we can move this forward without having the filename support the `storeKeysOnMachine` case?

I'd take a break, sleep over, and fix that :) Or at least mark that as "to be done".

**Comment by deepfire** [comment]
@domenkozar, I'm preparing a PR to facilitate specifying target location for the keys on top of this PR.

**Comment by domenkozar** [comment]
For supporting `storeKeysOnMachine` we would just do `keyFile = value.filename` and add an if/else to pick between text or filename.

@deepfire if that works, let's get it done :)

**Comment by deepfire** [comment]
I'm thinking of renaming this option to `keys.<key>.keyfile`, because of terminologic ambiguities arising -- what do you guys think, @domenkozar, @ip1981?

**Comment by domenkozar** [comment]
Maybe `source`? Then deployment location would be `target` and we can deprecate `text`.

**Comment by deepfire** [comment]
This would create an ambiguity -- `destination` ought to be a directory, right?  And `source` will be a file.

Type mismatch : -)

**Comment by deepfire** [comment]
Or, perhaps we can make `source` to be a directory too -- and fetch the key from there by its implied name, @domenkozar?

**Comment by domenkozar** [comment]
Good point. Let's then go with `keyFile` and `destination`.

**Comment by deepfire** [comment]
@domenkozar, just to make sure -- should we add the `destination` in this same PR (that's my current understanding)?

**Comment by domenkozar** [comment]
No, let's keep that separate.

**Comment by ip1981** [comment]
> `keys.<key>.keyfile`

I will love `keys.<key>.file`.

**Comment by deepfire** [comment]
@domenkozar, @ip1981, what do we decide about the option's name?

**Comment by ip1981** [comment]
I like shorter `file`, then `source`.

File: nix/keys.nix (hunk 16 above)
Comment (deepfire) [inline nix/keys.nix]: Supported, in the latest iteration.

**Comment by deepfire** [comment]
`storeKeysOnMachine = true` with `keyFile` supported in the latest iteration.

File: nix/keys.nix [hunk 17]
```diff
@@ -6,13 +6,38 @@ let
...
         <filename>/run/keys/<replaceable>password</replaceable></filename>
         will be <literal>foobar</literal>.
+
+        NOTE: Either <literal>text</literal> or <literal>keyFile</literal> have
+        to be set.
+      '';
+    };
+
+    options.keyFile = mkOption {
+      example = ./super-secret.key;
```
Comment (ip1981) [inline nix/keys.nix]: Ouch... I'm not sure this evaluates. Maybe drop example? Everyone knows what is a file :)

Comment (deepfire) [inline nix/keys.nix]: Hmm, oddly it never had given me a problem.  What was the error message?

But ok, removed it.

Comment (ip1981) [inline nix/keys.nix]: Nothing, really. Just random path in docs, maybe. Never mind.

**Comment by domenkozar** [comment]
I'm going to let this one here until tomorrow and merge.

File: nix/keys.nix [hunk 18]
```diff
@@ -123,32 +147,41 @@ in
...
         }
 
-        ${concatStringsSep "\n" (flip mapAttrsToList config.deployment.keys (name: value:
-          # Make sure each key has correct ownership, since the configured owning
-          # user or group may not have existed when first uploaded.
-          ''
-            [[ -f "/run/keys/${name}" ]] && chown '${value.user}:${value.group}' "/run/keys/${name}"
-          ''
-        ))}
+        ${optionalString (!config.deployment.storeKeysOnMachine)
```
Comment (domenkozar) [inline nix/keys.nix]: Why did you add this condition? Now  the code doesn't respect `user` and `group` settings anymore if I'm not mistaken.

Comment (deepfire) [inline nix/keys.nix]: @domenkozar, because calling `chown` on a symlink pointing to read-only Nix store failed, when I was testing it -- I was actually surprised how it worked before.

Comment (domenkozar) [inline nix/keys.nix]: Opened https://github.com/NixOS/nixops/issues/666

**Comment by domenkozar** [comment]
Besides #666 which was the case before this PR, I consider this done and working.
